core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceFactory.java

```diff
@@ -166,11 +170,26 @@ public static List<AnnotatedServiceElement> find(
                       .collect(toImmutableList());
     }
 
-    private static HttpStatus defaultResponseStatus(Method method) {
+    private static HttpStatus defaultResponseStatus(Method method, Class<?> clazz) {
         final StatusCode statusCodeAnnotation = AnnotationUtil.findFirst(method, StatusCode.class);
         if (statusCodeAnnotation == null) {
+            if (!producibleMediaTypes(method, clazz).isEmpty()) {
+                return HttpStatus.OK;
+            }
             // Set a default HTTP status code for a response depending on the return type of the method.
             final Class<?> returnType = method.getReturnType();
+
+            if (Publisher.class.isAssignableFrom(returnType) ||
+                CompletionStage.class.isAssignableFrom(returnType)) {
```

Comment:
Should we use `ClassUtil.unwrapAsyncType()`? `MONO_CLASS` seems enough for no_content status.

Reply:
I haven't noticed the existence of `ClassUtil.unwrapAsyncType(Type)` 👀.

From what I understand, `Publisher<Void>` indicates that you should not expect a data signal but rather only `onComplete()` and `onError(Throwable)` signals and it can cover other reactive implementations(e.g. [suspend fun to publisher](https://github.com/Kotlin/kotlinx.coroutines/blob/master/reactive/kotlinx-coroutines-reactive/src/Publish.kt#L36)), too so I thought it's reasonable to give 204 when the return type is `Publisher<Void>`.

And isn't it a bit weird to give different status codes depending on which return type is declared on the method?
```java
@Get("/foo") // 200
public Publisher<Void> voidPublisher() {
    return Mono.empty();
}

@Get("/bar") // 204
public Mono<Void> voidMono() {
    return Mono.empty();
}
```

Reply:
But, fixed the code anyways, to get aligned with the other code with regards to how they treat `Publisher` return type.

Reply:
That looks weird. 😆 
+1 for supporting `Publisher<Void>` as well.

Reply:
Makes sense. Let's return 204 for `Publisher<Void>`.

Reply:
Reverted the change but extracted the async type unwrapping logic here to `ClassUtil`, PTAL 🙏.